src/hwperfmon.c

```diff
@@ -0,0 +1,325 @@
+    case METAL_HWPERFMON_COUNTER_##x:                                          \
+        __asm__ __volatile__("csrr %0, mhpmevent" #x : "=r"(val));             \
+        val &= ~bitmask;                                                       \
+        val |= bitmask;                                                        \
+        __asm__ __volatile__("csrw mhpmevent" #x ", %0" : : "r"(val));         \
+        break;
+
+/* Macro to set values into event selector register */
+#define METAL_HWPERFMON_CLR_EVENT_REG(x)                                       \
+    case METAL_HWPERFMON_COUNTER_##x:                                          \
+        __asm__ __volatile__("csrr %0, mhpmevent" #x : "=r"(val));             \
+        val &= ~bitmask;                                                       \
+        __asm__ __volatile__("csrw mhpmevent" #x ", %0" : : "r"(val));         \
+        break;
+
+/* Macro to get values from event selector register */
+#define METAL_HWPERFMON_GET_EVENT_REG(x)                                       \
+    case METAL_HWPERFMON_COUNTER_##x:                                          \
+        __asm__ __volatile__("csrr %0, mhpmevent" #x : "=r"(val));             \
+        break;
+
+/* Macro to read HW performance monitor counter values */
+#if __riscv_xlen == 32
+#define METAL_HWPERFMON_GET_COUNT_REG(x)                                       \
+    case METAL_HWPERFMON_COUNTER_##x:                                          \
+        __asm__ __volatile__("csrr %0, mhpmcounter" #x "h" : "=r"(vh));        \
+        __asm__ __volatile__("csrr %0, mhpmcounter" #x : "=r"(vl));            \
+        break;
+#else
+#define METAL_HWPERFMON_GET_COUNT_REG(x)                                       \
+    case METAL_HWPERFMON_COUNTER_##x:                                          \
+        __asm__ __volatile__("csrr %0, mhpmcounter" #x : "=r"(vl));            \
+        break;
+#endif
+
+/* Macro to clear HW performance monitor counter values */
+#if __riscv_xlen == 32
+#define METAL_HWPERFMON_CLR_COUNT_REG(x)                                       \
+    case METAL_HWPERFMON_COUNTER_##x:                                          \
+        __asm__ __volatile__("csrw mhpmcounter" #x "h, zero");                 \
+        __asm__ __volatile__("csrw mhpmcounter" #x ", zero");                  \
+        break;
+#else
+#define METAL_HWPERFMON_CLR_COUNT_REG(x)                                       \
+    case METAL_HWPERFMON_COUNTER_##x:                                          \
+        __asm__ __volatile__("csrw mhpmcounter" #x ", zero");                  \
+        break;
+#endif
+
+/* Macro to check for instruction trap */
+#define MCAUSE_ILLEGAL_INST 0x02
+
+/* Return codes */
+#define METAL_HWPERFMON_RET_OK 0
+#define METAL_HWPERFMON_RET_NOK 1
+
+/* Module enable flag */
+static int hwperfmon_enable = 0;
+
+/* Available hardware counters */
+static metal_hwperfmon_counter hwperfmon_count = 0;

```

Comment:
These global variable won't work for MC designs.

Reply:
Any per-hart storage should be generated by `freedom-metal_header-generator`, in this case probably as members of the RISC-V cpu struct.

Reply:
Added checks to make sure only single hart now writes to global variables.

Reply:
How will that help? Are the hwperf enable bits shared by all harts?

Reply:
My bad, i mistakenly assumed that these counters are shared between harts. hwperfmon_count is now moved into metal cpu device instance, as suggested by Nate.